File: app/src/main/java/com/saitejajanjirala/imdbclone/ui/bookmark/BookMarkViewModel.java
package com.saitejajanjirala.imdbclone.ui.bookmark;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.saitejajanjirala.imdbclone.models.Result;
import com.saitejajanjirala.imdbclone.repo.IMDBRepo;
import com.saitejajanjirala.imdbclone.listeners.BookMarkListener;

import java.util.List;

public class BookMarkViewModel extends ViewModel {
    private IMDBRepo imdbRepo;
    MutableLiveData<List<Result>> bookMarkResults = new MutableLiveData<>();

    public BookMarkViewModel(IMDBRepo imdbRepo) {
        this.imdbRepo = imdbRepo;
        getData();
    }
    void getData(){
        imdbRepo.getBookMarkedResults(bookMarkResults);
    }

    void setBookMarkStatus(Result result, int position, BookMarkListener listener) {
        imdbRepo.setBookMarkingStatus(result, position, listener);
    }

    void clearBookMarks(){
        imdbRepo.clearBookMarked(bookMarkResults);
    }


}

File: app/src/main/java/com/saitejajanjirala/imdbclonekotlin/ui/bookmark/BookMarkViewModel.kt
package com.saitejajanjirala.imdbclonekotlin.ui.bookmark

import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.saitejajanjirala.imdbclonekotlin.db.DatabaseService
import com.saitejajanjirala.imdbclonekotlin.listeners.BookMarkListener
import com.saitejajanjirala.imdbclonekotlin.models.Result
import com.saitejajanjirala.imdbclonekotlin.network.ApiService
import com.saitejajanjirala.imdbclonekotlin.utils.Helper
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class BookMarkViewModel @Inject constructor(
    private val apiService: ApiService,
    private val helper: Helper,
    private val databaseService: DatabaseService
) : ViewModel() {
    val bookMarkResults=MutableLiveData<List<Result>>()
    val msg=MutableLiveData<String>()
    fun clearBookMarks() {
        viewModelScope.launch {
            try {
                coroutineScope {
                    try {
                        val resultDao = databaseService.resultDao()
                        resultDao.clearBookMarks()
                        getData()
                    } catch (e: Exception) {
                        msg.postValue("something went wrong")
                    }
                }
            }
            catch (e:Exception){
                Log.i("clear bookmarks error",e.message.toString())
                msg.postValue("something went wrong")
            }
        }
    }

    init {
        getData()
    }

    private fun getData() {
        viewModelScope.launch {
            try{
                val results=databaseService.resultDao().bookMarkedResults()
                Log.i("bookmark results",results.toString())
                bookMarkResults.postValue(results)
            }
            catch (e:Exception){
                Log.i("get data error bookmark",e.message.toString())
            }
        }
    }

    fun setBookMarkStatus(result: Result?, position: Int, listener: BookMarkListener?) {
        viewModelScope.launch {
            try{
               if(result!=null && listener!=null){
                   result.bookmarked=!result.bookmarked
                   databaseService.resultDao().updateBookMarked(result)
                   listener.onBookMarkClicked(position,result)
               }
            }
            catch (e:Exception){
                msg.postValue("Something went wrong")
                Log.i("get data error bookmark",e.message.toString())
            }
        }
    }


}